orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/providers/gce/GceDeployStagePreProcessor.java

```diff
@@ -0,0 +1,152 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class GceDeployStagePreProcessor implements DeployStagePreProcessor {
+  private final ApplySourceServerGroupCapacityStage applySourceServerGroupSnapshotStage;
+  private final ResizeServerGroupStage resizeServerGroupStage;
+  private final TargetServerGroupResolver targetServerGroupResolver;
+
+  @Autowired
+  public GceDeployStagePreProcessor(
+      ApplySourceServerGroupCapacityStage applySourceServerGroupCapacityStage,
+      ResizeServerGroupStage resizeServerGroupStage,
+      TargetServerGroupResolver targetServerGroupResolver) {
+    this.applySourceServerGroupSnapshotStage = applySourceServerGroupCapacityStage;
+    this.resizeServerGroupStage = resizeServerGroupStage;
+    this.targetServerGroupResolver = targetServerGroupResolver;
+  }
+
+  @Override
+  public List<StepDefinition> additionalSteps(Stage stage) {
+    return ImmutableList.of(
+        new StepDefinition(
+            "snapshotSourceServerGroup", CaptureSourceServerGroupCapacityTask.class));
+  }
+
+  @Override
+  public List<StageDefinition> beforeStageDefinitions(Stage stage) {
+    StageData stageData = stage.mapTo(StageData.class);
+    if (shouldPinSourceServerGroup(stageData.getStrategy())) {
+      Map<String, Object> resizeContext = getResizeContext(stageData);
+      resizeContext.put("pinMinimumCapacity", true);
+
+      return ImmutableList.of(
+          new StageDefinition(
+              String.format("Pin %s", resizeContext.get("serverGroupName")),
+              resizeServerGroupStage,
+              resizeContext));
+    }
+
+    return ImmutableList.of();
+  }
+
+  @Override
+  public List<StageDefinition> afterStageDefinitions(Stage stage) {
+    return ImmutableList.of(
+        new StageDefinition(
+            "restoreMinCapacityFromSnapshot",
+            applySourceServerGroupSnapshotStage,
+            Collections.emptyMap()));
+  }
+
+  @Override
+  public List<StageDefinition> onFailureStageDefinitions(Stage stage) {
+    StageData stageData = stage.mapTo(StageData.class);
+
+    StageDefinition unpinServerGroupStage = buildUnpinServerGroupStage(stageData);
+    if (unpinServerGroupStage == null) {
+      return ImmutableList.of();
+    }
+
+    return ImmutableList.of(unpinServerGroupStage);
+  }
+
+  @Override
+  public boolean supports(Stage stage) {
+    StageData stageData = stage.mapTo(StageData.class);
+    return stageData.getCloudProvider().equals("gce");
+  }
+
+  private static boolean shouldPinSourceServerGroup(String strategy) {
+    return Strategy.fromStrategyKey(strategy) == Strategy.RED_BLACK;
+  }
+
+  private Map<String, Object> getResizeContext(StageData stageData) {
+    AbstractDeployStrategyStage.CleanupConfig cleanupConfig =
+        AbstractDeployStrategyStage.CleanupConfig.fromStage(stageData);
+    Map<String, Object> resizeContext = new HashMap<>();
+
+    resizeContext.put(
```

Comment:
Great idea, I didn't even notice that! Since the next PR in this thrilling series is going to change even more about `getResizeContext`, I'll include the addition of a `toContext` method as a commit with that PR!